vllm/model_executor/models/qwen.py

```diff
@@ -237,16 +741,242 @@ def forward(
         return hidden_states
 
 
-class QWenLMHeadModel(nn.Module):
+def get_image_text(image_num: int, padding: bool) -> str:
+    """Retrieves a placeholder text that when tokenized, will be expanded with
+    image pads.
+
+    Args:
+        image_num: The number of the image that we want a text prompt for.
+            Images should be indexed starting at 1.
+        padding: Whether or not padding should be manually added.
+
+    Returns:
+        Text placeholder prompt for the image being considered.
+    """
+    image_start = f"Picture {image_num}: {IMG_START}"
+    image_end = f"{IMG_END}\n"
+    if not padding:
+        return f"{image_start}{image_end}"
+    return f"{image_start}{MAX_QWEN_IMG_TOKENS * IMG_PAD}{image_end}"
+
+
+def input_processor_for_qwen(ctx: InputContext, llm_inputs: LLMInputs):
+    """Processes the inputs, which may or may not be multimodal.
+    Multimodal inputs will only be processed if the model has a "visual"
+    component in its model config, otherwise they'll be ignored.
+
+    Args:
+        ctx: Context of the loaded model.
+        llm_inputs: LLM inputs which may have a multi_modal_data attribute.
+
+    Returns:
+        If the model is language only or not multimodal inputs were provided,
+        returns llm_inputs unmodified. Otherwise, processes the multimodal
+        images / image embeddings and adds the fixed-length image placeholders.
+    """
+    multi_modal_data = llm_inputs.get("multi_modal_data")
+
+    # Only process images if we have multimodal data and a visual config
+    hf_config = ctx.get_hf_config()
+    if (multi_modal_data is None or "image" not in multi_modal_data
+            or not hasattr(hf_config, "visual")):
+        return llm_inputs
+
+    prompt = llm_inputs.get("prompt")
+    prompt_token_ids = llm_inputs["prompt_token_ids"]
+    model_config = ctx.model_config
+    tokenizer = cached_get_tokenizer(model_config.tokenizer,
+                                     trust_remote_code=True)
+    image_data = multi_modal_data["image"]
+    if isinstance(image_data, torch.Tensor):
+        num_dims = len(image_data.shape)
+        if num_dims < 2 or num_dims > 3:
+            raise ValueError(
+                f"Expected img embeds to be have 3 dimensions, got {num_dims}")
+        num_images = 1 if num_dims == 2 else image_data.shape[0]
+    else:
+        # TODO - handle multiple image inputs once the API is solidified
+        num_images = 1
+
+    if prompt is None:
+        prompt = tokenizer.decode(prompt_token_ids)
+
+    # Iteratively replace image tags for every image that we expect
+    # Currently we only allow multiple images input as embeddings.
+    num_img_tags = prompt.count("<image>")
```

Comment:
This is a bit of a point of awkwardness that it would be cool to discuss if anyone has thoughts. The Qwen tokenizer has a utility func called `from_list_format` which handles encoding images something like the following.

```
>>> tokenizer.from_list_format([
...     {'image': 'https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg'},
...     {'text': '这是什么'},
... ])
'Picture 1: <img>https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg</img>\n这是什么'
```

When the model is invoked, it searches between `<img> / </img>` to check if it's a URL or local path, then takes the appropriate action to get it into a PIL image. Currently, the examples are written against just taking an `<image>` tag and getting the corresponding image prompt to avoid potentially messing up numbering etc, since making mistakes, e.g., 0-indexing the pictures, can do some weird stuff to results, and this model does support multiple images, which is currently available through embeddings.

I am working on fixing the image processor to better align with the format for the hf model, but was wondering if people had any thoughts on potentially:

- Supporting the format created through `from_list_format` for images, with the caveat that URLs won't be resolved
- Additionally supporting just using `<image>` as is and allowing for automatic substitution of the image prompts if used - I feel people may make mistakes in this format on the client side unless they are using the Qwen tokenizer utils to form their query